# ãƒ‡ã‚£ãƒ¬ã‚¯ã‚¿ãƒ¼ï¼ˆDirectorï¼›ç›£ç£ï¼‰
#	ã¨ã‚Šã‚ãˆãšã€ã‚²ãƒ¼ãƒ å…¨ä½“ã‚’ã‚³ãƒ³ãƒˆãƒ­ãƒ¼ãƒ«ã™ã‚‹
extends Node2D


var DepartmentSnapshot = load("res://scripts/department_snapshot.gd")


# ç¾åœ¨ã®éƒ¨é–€ï¼ˆStringNameå‹ï¼‰
var current_department_name = null
# ç¾åœ¨é³´ã£ã¦ã„ã‚‹èƒŒæ™¯éŸ³æ¥½ã®ãƒãƒ¼ãƒ‰å
var current_bgm_name = null
# ç¾åœ¨é³´ã£ã¦ã„ã‚‹åŠ¹æœéŸ³ã®ãƒãƒ¼ãƒ‰å
var current_se_name = null
# ã‚¹ãƒŠãƒƒãƒ—ã‚·ãƒ§ãƒƒãƒˆè¾æ›¸ï¼ˆã‚­ãƒ¼ï¼šStringNameå‹ï¼‰
var snapshots = {}
# ãƒˆæ›¸ãï¼ˆã‚·ãƒŠãƒªã‚ªã®å‘½ä»¤ãƒ‘ãƒ©ã‚°ãƒ©ãƒ•ï¼‰ã§ä½¿ãˆã‚‹å¤‰æ•°ã®è¾æ›¸
var stage_directions_variables = {}
# ç–‘ä¼¼çš„ãªã‚¹ãƒªãƒ¼ãƒ—ã«ä½¿ã†ã‚«ã‚¦ãƒ³ãƒˆ
var sleep_seconds = 0.0


# åŠ©ç›£å–å¾—
func get_assistant_director():
	return $"AssistantDirector"


# éƒ¨é–€åˆ‡æ›¿å–å¾—
func get_switch_department():
	return $"ScenarioWriter/SwitchDepartment"


# ã‚¹ãƒŠãƒƒãƒ—ã‚·ãƒ§ãƒƒãƒˆ
func get_snapshot(
	department_name):	# StringName
	return self.snapshots[department_name]


func get_current_snapshot():
	return self.get_snapshot(self.current_department_name)


# ä¼è¨€çª“ï¼ˆç¾åœ¨ã€å‡ºåŠ›ã®å¯¾è±¡ã«ãªã£ã¦ã„ã‚‹ã‚‚ã®ï¼‰
func get_message_window(node_name_obj):
	print("ï¼»ç›£ç£ï¼½ã€€ä¼è¨€çª“åï¼šï¼»" + str(node_name_obj) + "ï¼½")
	return $"GuiArtist/WindowsOfMessage".get_node(str(node_name_obj))


# ä¼è¨€çª“ï¼ˆç¾åœ¨ã€å‡ºåŠ›ã®å¯¾è±¡ã«ãªã£ã¦ã„ã‚‹ã‚‚ã®ï¼‰
func get_current_message_window():
	var snapshot = self.get_current_snapshot()
	if snapshot.stack_of_last_displayed_message_window.size() < 1:
		print("ï¼»ç›£ç£ï¼½ã€€â–²ï¼ã€€æœ€å¾Œã«è¡¨ç¤ºã—ãŸãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã‚¦ã‚£ãƒ³ãƒ‰ã‚¦ãŒç„¡ã„")

	var node_name = snapshot.stack_of_last_displayed_message_window[-1]
	print("ï¼»ç›£ç£ï¼½ã€€ä¼è¨€çª“åï¼šï¼»" + node_name + "ï¼½")
	return self.get_message_window(str(node_name))


# ç¾åœ¨ã®ã€ŒÂ§ã€ã‚»ã‚¯ã‚·ãƒ§ãƒ³è¨­å®š
func set_current_section(section_name):
	var snapshot = self.get_current_snapshot()
	snapshot.section_name = section_name
	snapshot.section_item_index = 0


# å…¨ã¦ã®éƒ¨é–€åä¸€è¦§
func get_all_department_names():
	var array = []	# StringName ã®é…åˆ—
	
	for department in $"ScenarioWriter".get_children():
		# SwitchDepartment ã¨ System ã¯é™¤ã
		if department.name != "SwitchDepartment" and department.name != "System":
			array.append(department.name)

	return array


# å„éƒ¨é–€ãŒæœ€å¾Œã«é–‹ã„ã¦ã„ãŸãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ãƒ»ã‚¦ã‚£ãƒ³ãƒ‰ã‚¦åã®ä¸€è¦§ã‚’è¡¨ç¤º
func dump_last_displayed_message_window():
	print("ï¼»ç›£ç£ï¼½ã€€å„éƒ¨é–€ãŒæœ€å¾Œã«é–‹ã„ã¦ã„ãŸãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ãƒ»ã‚¦ã‚£ãƒ³ãƒ‰ã‚¦åã®ä¸€è¦§ã‚’è¡¨ç¤º")
	
	# éƒ¨é–€åä¸€è¦§
	var department_names = self.get_all_department_names()
	for department_name in 	department_names:
		print("ã€€ã€€éƒ¨é–€ï¼šã€€" + department_name)

		# ã‚¹ãƒŠãƒƒãƒ—ã‚·ãƒ§ãƒƒãƒˆ
		var snapshot = self.get_snapshot(department_name)
		
		for window_name in snapshot.node_names_of_currently_displayed_message_window:
			print("ã€€ã€€ã€€ã€€ğŸ‘ " + window_name)


# ã‚µãƒ–ãƒ„ãƒªãƒ¼ãŒå…¨ã¦ã‚¤ãƒ³ã‚¹ã‚¿ãƒ³ã‚¹åŒ–ã•ã‚ŒãŸã¨ãã«å‘¼ã³å‡ºã•ã‚Œã‚‹
# Called when the node enters the scene tree for the first time.
func _ready():

	# ã‚¹ãƒŠãƒƒãƒ—ã‚·ãƒ§ãƒƒãƒˆè¾æ›¸ä½œæˆ
	for department_name in self.get_all_department_names():
		var department_node = $"ScenarioWriter".get_node(str(department_name))
		if department_node.name != "SwitchDepartment" and department_node.name != "System":
			self.snapshots[department_node.name] = DepartmentSnapshot.new()

			# ï¼ˆã‚ã‚“ã©ãã•ã„ã‘ã©ï¼‰ SwitchDepartment ã‹ã‚‰ãƒ—ãƒ­ãƒ‘ãƒ†ã‚£ã‚’ç§»ã™
			self.snapshots[department_node.name].name = department_node.name		# StringName å‹

			# TODO ã“ã®åˆæœŸåŒ–ã¯å¿…è¦ã‹ï¼Ÿ
			# ãƒ¡ãƒƒã‚»ãƒ¼ã‚¸ã‚’å‡ºåŠ›ã™ã‚‹å¯¾è±¡ã¨ãªã‚‹ã‚¦ã‚£ãƒ³ãƒ‰ã‚¦ã®åå‰ï¼ˆæ–‡å­—åˆ—ï¼‰ã€‚ãƒŒãƒ«ã«ã›ãšã€å¿…ãšä½•ã‹å…¥ã‚Œã¦ãŠã„ãŸæ–¹ãŒãƒ‡ãƒãƒƒã‚°ã—ã‚„ã™ã„
			if department_node.name =="ğŸ“—ãƒ“ã‚¸ãƒ¥ã‚¢ãƒ«ãƒãƒ™ãƒ«éƒ¨é–€":
				self.snapshots[department_node.name].stack_of_last_displayed_message_window.push_back(&"â– ä¸‹")	# StringName å‹ ã‚·ãƒ³ã‚¿ãƒƒã‚¯ã‚¹ãƒ»ã‚·ãƒ¥ã‚¬ãƒ¼
			elif department_node.name =="ğŸ“—ã‚·ã‚¹ãƒ†ãƒ ãƒ¡ãƒ‹ãƒ¥ãƒ¼éƒ¨é–€":
				self.snapshots[department_node.name].stack_of_last_displayed_message_window.push_back(&"â– ä¸­å¤®")	# StringName å‹ ã‚·ãƒ³ã‚¿ãƒƒã‚¯ã‚¹ãƒ»ã‚·ãƒ¥ã‚¬ãƒ¼
				#self.snapshots[department_node.name].stack_of_last_displayed_message_window.push_back(&"â– å·¦ä¸‹")	# StringName å‹ ã‚·ãƒ³ã‚¿ãƒƒã‚¯ã‚¹ãƒ»ã‚·ãƒ¥ã‚¬ãƒ¼
			elif department_node.name =="ğŸ“—ãƒãƒˆãƒ«éƒ¨é–€":
				self.snapshots[department_node.name].stack_of_last_displayed_message_window.push_back(&"â– ä¸‹")	# StringName å‹ ã‚·ãƒ³ã‚¿ãƒƒã‚¯ã‚¹ãƒ»ã‚·ãƒ¥ã‚¬ãƒ¼


			# æ–‡æ›¸è¾æ›¸ã®å…ˆé ­è¦ç´ ã®ã‚­ãƒ¼å–å¾—
			self.snapshots[department_node.name].section_name = $"ScenarioWriter".get_node(str(department_node.name)).document.keys()[0]

	# é–‹ç™ºä¸­ã«ã„ã˜ã£ãŸã‚‚ã®ãŒæ®‹ã£ã¦ã‚‹ã‹ã‚‚ã—ã‚Œãªã„ã‹ã‚‰ã€æƒé™¤
	#
	# ç›£ç£ã¯è¡¨ç¤ºã™ã‚‹å¿…è¦ãŒã‚ã‚‹
	self.show()
	# ï¼§ï¼µï¼©ã‚¢ãƒ¼ãƒ†ã‚£ã‚¹ãƒˆè‡ªèº«ã¯è¡¨ç¤ºã™ã‚‹å¿…è¦ãŒã‚ã‚‹
	$"GuiArtist".show()
	# ãƒ¢ãƒ³ã‚¹ã‚¿ãƒ¼ãƒ»ãƒˆãƒ¬ãƒ¼ãƒŠãƒ¼
	$"MonsterTrainer".show()
	$"MonsterTrainer/Faces".show()
	$"MonsterTrainer/WholeBody".show()
	
	# ã‚°ãƒªãƒƒãƒ‰ã¯éš ã™
	$"Grid".hide()

	# èƒŒæ™¯ç”»åƒã¯å…¨éƒ¨éš ã™
	for sprite2d_node in $"BackgroundArtist".get_children():
		sprite2d_node.hide()

	# ã‚¦ã‚£ãƒ³ãƒ‰ã‚¦ã¯ã¨ã«ã‹ãéš ã™
	#
	#	è¦ªãƒãƒ¼ãƒ‰ã¯ä¾‹å¤–
	$"GuiArtist/WindowsOfMessage".show()
	#	ä¼è¨€çª“ã¯ã¨ã«ã‹ãéš ã™
	for message_window in $"GuiArtist/WindowsOfMessage".get_children():
		message_window.hide()
	#
	#	ãƒ“ãƒ¥ãƒ¼ã‚¤ãƒ³ã‚°ãƒ»ã‚¦ã‚£ãƒ³ãƒ‰ã‚¦ã¯ã¨ã«ã‹ãéš ã™
	for sprite2d_node in $"GuiArtist/WindowsOfViewing".get_children():
		sprite2d_node.hide()
	$"GuiArtist/WindowsOfViewing/System/Frame".hide()
	#
	#	ãƒ†ãƒ­ãƒƒãƒ—ã¯ã¨ã«ã‹ãéè¡¨ç¤ºã«ã™ã‚‹
	for canvas_layer in $"TelopCoordinator".get_children():
		canvas_layer.hide()
	#
	#	ãƒ¢ãƒ³ã‚¹ã‚¿ãƒ¼ã¯ã€ãƒ•ã‚©ãƒ«ãƒ€â€•ä»¥å¤–ã¯ã€€ã¨ã«ã‹ãéè¡¨ç¤ºã«ã™ã‚‹
	for monster in $"MonsterTrainer/Faces".get_children():
		monster.hide()
	for monster in $"MonsterTrainer/WholeBody".get_children():
		monster.hide()

	#

	# æœ€åˆã«å®Ÿè¡Œã™ã‚‹éƒ¨é–€å
	self.current_department_name = self.get_switch_department().start_department_name

	var snapshot = self.get_current_snapshot()

	# ãƒ‘ãƒ¼ã‚¹ã™ã‚‹ãª
	snapshot.set_parse_lock(true)

	# å°æœ¬ã®ã€ŒÂ§ã€ã‚»ã‚¯ã‚·ãƒ§ãƒ³ã®å†ç”Ÿ
	$"./AssistantDirector".play_section()

	# ä¼è¨€çª“ã‚’ã€ä¸€æ™‚çš„ã«å±…ãªããªã£ã¦ã„ãŸã®ã‚’è§£é™¤ã™ã‚‹
	self.get_current_message_window().set_appear_subtree(true)


# ãƒ†ã‚­ã‚¹ãƒˆãƒœãƒƒã‚¯ã‚¹ãªã©ã«ãƒ•ã‚©ãƒ¼ã‚«ã‚¹ãŒç„¡ã„ã¨ãã®å…¥åŠ›ã‚’æ‹¾ã†
#
# å­è¦ç´ ã‹ã‚‰è¦ªè¦ç´ ã®é †ã§å‘¼ã³å‡ºã•ã‚Œã‚‹ã‚ˆã†ã ã€‚
# ã“ã®ãƒ—ãƒ­ã‚°ãƒ©ãƒ ã§ã¯ã€€ãƒ«ãƒ¼ãƒˆã€€ã ã‘ã§ã€€ã‚­ãƒ¼å…¥åŠ›ã‚’æ‹¾ã†ã“ã¨ã«ã™ã‚‹
func _unhandled_key_input(event):

	# ä½•ã‹ã‚­ãƒ¼ã‚’æŠ¼ã—ãŸã¨ã
	if event.is_pressed():

		# ã“ã®ã‚²ãƒ¼ãƒ ç‹¬è‡ªã®ä»®æƒ³ã‚­ãƒ¼ã«å¤‰æ›
		var virtual_key = ""
		
		# ã‚¨ãƒ³ã‚¿ãƒ¼ã‚­ãƒ¼æŠ¼ä¸‹
		if event.keycode == KEY_ENTER:
			virtual_key = &"VK_Ok"

		# ã‚¨ã‚¹ã‚±ãƒ¼ãƒ—ã‚­ãƒ¼æŠ¼ä¸‹
		elif event.keycode == KEY_ESCAPE:
			virtual_key = &"VK_SystemMenu"

		# ï¼»ï¼²ï¼½ã‚­ãƒ¼æŠ¼ä¸‹ï¼ˆå¾Œã§ã‚¹ãƒ¼ãƒ‘ãƒ¼ãƒ•ã‚¡ãƒŸã‚³ãƒ³ã® R ã‚­ãƒ¼ã«ã—ã‚ˆã†ã¨æ€ã£ã¦ã„ãŸã‚¢ãƒ«ãƒ•ã‚¡ãƒ™ãƒƒãƒˆï¼‰
		elif event.keycode == KEY_R:
			virtual_key = &"VK_FastForward"

		# ä»®æƒ³ã‚­ãƒ¼ã‚’æŠ¼ä¸‹ã—ãŸã¨ã„ã†å»ºä»˜ã‘
		self.on_virtual_key_pressed(virtual_key)
		
	else:
		print("ï¼»ç›£ç£ï¼½ã€€ã‚¢ãƒ³ãƒãƒ³ãƒ‰ãƒ«ãƒ‰ãƒ»ã‚­ãƒ¼ã€€æŠ¼ä¸‹ä»¥å¤–")

func _unhandled_input(event):

	# ã“ã®ã‚²ãƒ¼ãƒ ç‹¬è‡ªã®ä»®æƒ³ã‚­ãƒ¼ã«å¤‰æ›
	var virtual_key = &""
	
	# æ–‡å­—åˆ—ã ã‘ã ã¨ã€æŠ¼ã—ãŸã®ã‹æ”¾ã—ãŸã®ã‹åˆ†ã‹ã‚‰ãªã„
	var event_as_text = event.as_text()
	
	if event.is_pressed():
		
		# ã‚ªãƒ¼ã‚±ãƒ¼ç›¸å½“ã®ãƒœã‚¿ãƒ³æŠ¼ä¸‹
		if event_as_text == &"Joypad Button 0 (Bottom Action, Sony Cross, Xbox A, Nintendo B)":
			virtual_key = &"VK_Ok"

		# ã‚¹ã‚¿ãƒ¼ãƒˆãƒœã‚¿ãƒ³æŠ¼ä¸‹
		elif event_as_text == &"Joypad Button 4 (Back, Sony Select, Xbox Back, Nintendo -)":
			virtual_key = &"VK_SystemMenu"
		
		# PC-Engine ã®ã‚²ãƒ¼ãƒ ãƒ‘ãƒƒãƒ‰ã§ã¯ã€ãƒšãƒ¼ã‚¸æ—©é€ã‚Šã®æ©Ÿèƒ½ã‚’æŒãŸã›ã‚‹ãƒœã‚¿ãƒ³ãŒè¶³ã‚Šãªã„ã€‚ã‚­ãƒ¼ãƒœãƒ¼ãƒ‰ã‚’ä½µç”¨ã—ã¦ã‚‚ã‚‰ã†ã“ã¨

		# ä»®æƒ³ã‚­ãƒ¼ã‚’æŠ¼ä¸‹ã—ãŸã¨ã„ã†å»ºä»˜ã‘
		self.on_virtual_key_pressed(virtual_key)


# ä»®æƒ³ã‚­ãƒ¼ã‚’æŠ¼ä¸‹ã—ãŸã¨ã„ã†å»ºä»˜ã‘
func on_virtual_key_pressed(virtual_key):
	# ç¾åœ¨ã®ãƒ‡ãƒ‘ãƒ¼ãƒˆãƒ¡ãƒ³ãƒˆã«ç´ã¥ãã€é …ç›®ã¯è¾æ›¸ã«è¨˜è¼‰ã•ã‚Œã¦ã„ã‚‹ã‹ï¼Ÿ
	if str(self.current_department_name) in self.get_switch_department().key_pressed_stage_directions:
		
		# ãã®è¦ç´ ã‚’å–å¾—
		var key_pressed_stage_directions_1 = self.get_switch_department().key_pressed_stage_directions[str(self.current_department_name)]
		
		# æŠ¼ã—ãŸã‚­ãƒ¼ã«ç´ã¥ãã€ãƒˆæ›¸ãã¯è¾æ›¸ã«è¨˜è¼‰ã•ã‚Œã¦ã„ã‚‹ã‹ï¼Ÿ
		if virtual_key in key_pressed_stage_directions_1:
			
			# ãã®ãƒˆæ›¸ã
			var stage_directions = key_pressed_stage_directions_1[virtual_key]

			print("ï¼»ç›£ç£ï¼½ã€€ã‚¢ãƒ³ãƒãƒ³ãƒ‰ãƒ«ãƒ‰ãƒ»ã‚­ãƒ¼æŠ¼ä¸‹ã€€éƒ¨é–€å¤‰æ›´")

			# TODO ã“ã“ã§ stage_directions ã‚’ãƒˆæ›¸ãã¨ã—ã¦å®Ÿè¡Œã—ãŸã„ãŒã€ã§ãã‚‹ã‹ï¼Ÿ
			self.get_assistant_director().parse_paragraph(stage_directions)

			# å­è¦ç´ ã«ã¯æ¸¡ã—ã¾ã›ã‚“
			return

	print("ï¼»ç›£ç£ï¼½ã€€ã‚¢ãƒ³ãƒãƒ³ãƒ‰ãƒ«ãƒ‰ãƒ»ã‚­ãƒ¼æŠ¼ä¸‹ã€€ãã®ä»–ã®ã‚­ãƒ¼")

	#	å­è¦ç´ ã¸æ¸¡ã™
	self.get_current_message_window().on_unhandled_virtual_key_input(virtual_key, &"VKO_Pressed")
